### PR TITLE
Add `schema` parameter to `read_parquet`

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR implements the `schema` parameter for `read_parquet`, which allows us to read a Parquet file _as if_ it has the supplied schema. Field IDs are required. For example:
```sql
COPY (SELECT 42::INTEGER i) TO 'integers.parquet' (FIELD_IDS {i: 0});
SELECT *
FROM read_parquet('integers.parquet', schema=map {
                    0: {name: 'renamed_i', type: 'BIGINT', default_value: NULL},
                    1: {name: 'new_column', type: 'UTINYINT', default_value: 43}
                  });
-- ┌───────────┬────────────┐
-- │ renamed_i │ new_column │
-- │   int64   │   uint8    │
-- ├───────────┼────────────┤
-- │        42 │         43 │
-- └───────────┴────────────┘
```
Columns are identified by field id and can be added, deleted, reordered, renamed, and cast to a different type.

This parameter cannot be combined with `union_by_name=true`, and for now, it also cannot be combined with hive partitioning. Nested types are also not (yet) supported.

### CODE REVIEW

File: extension/parquet/parquet_extension.cpp [hunk 1]
```diff
@@ -157,6 +159,89 @@ BindInfo ParquetGetBatchInfo(const FunctionData *bind_data) {
 	return bind_info;
 }
 
+static MultiFileReaderBindData BindSchema(ClientContext &context, vector<LogicalType> &return_types,
+                                          vector<string> &names, ParquetReadBindData &result, ParquetOptions &options) {
+	D_ASSERT(!options.schema.empty());
+	auto &file_options = options.file_options;
+	if (file_options.union_by_name || file_options.hive_partitioning) {
```
Comment (samansmink) [inline extension/parquet/parquet_extension.cpp:166]: Would be nice to throw a separate error for when hive auto detection is enabled. In the following case:
```
statement ok
COPY (
    SELECT 1 i1, 3 i3, 4 i4, 5 i5 UNION ALL
    SELECT 2 i1, 3 i3, 4 i4, 5 i5
) TO '__TEST_DIR__/partitioned' (FIELD_IDS {i1: 5, i3: 3, i4: 2, i5: 1}, PARTITION_BY i1, FORMAT parquet)

query IIII
SELECT *
FROM read_parquet('__TEST_DIR__/partitioned/*/*.parquet', schema=map {
                    1: {name: 'i1', type: 'BIGINT', default_value: NULL},
                    3: {name: 'i3', type: 'BIGINT', default_value: NULL},
                    4: {name: 'i4', type: 'BIGINT', default_value: 2},
                    5: {name: 'i5', type: 'BIGINT', default_value: NULL}
                  })
----
5	3	2	1
5	3	2	2
```

The binder error is thrown but hive_partitioning is not explicitly enabled.

File: test/parquet/test_parquet_schema.test [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    SELECT 1 i1, 3 i3, 4 i4, 5 i5
+) TO '__TEST_DIR__/integers.parquet' (FIELD_IDS {i1: 5, i3: 3, i4: 2, i5: 1})
+
+# this is purposely a bit confusing but we're:
+# 1. deleting field id 2
+# 2. creating field id 4
+# 3. reversing the order of the columns
+# 4. renaming them (except i3)
+# 5. upcasting them
+query IIII
```
Comment (samansmink) [inline test/parquet/test_parquet_schema.test:143]: Maybe we can expand/add a test here that tests that it also works with multiple parquet files (but without hive partitioning)?
something like: 
```
statement ok
COPY (
    SELECT 1 i1, 3 i3, 4 i4, 5 i5
) TO '__TEST_DIR__/multifile1.parquet' (FIELD_IDS {i1: 5, i3: 3, i4: 2, i5: 1})

statement ok
COPY (
    SELECT 1 i1, 3 i3, 4 i4, 5 i5
) TO '__TEST_DIR__/multifile2.parquet' (FIELD_IDS {i1: 1, i3: 2, i4: 3, i5: 4})

query IIII
SELECT *
FROM read_parquet('__TEST_DIR__/multifile*.parquet', schema=map {
                    1: {name: 'i1', type: 'BIGINT', default_value: NULL},
                    3: {name: 'i3', type: 'BIGINT', default_value: NULL},
                    4: {name: 'i4', type: 'BIGINT', default_value: 2},
                    5: {name: 'i5', type: 'BIGINT', default_value: NULL}
                  })
ORDER BY i1
----
1	4	5	NULL
5	3	2	1
```

**Review by samansmink — COMMENTED** [review]
Very cool! I added 2 minor comments, but in general it looks good!

**Comment by lnkuiper** [comment]
Thanks for the feedback! I've added the tests, and disabled auto-detection of hive partitioning when the schema parameter is used

**Comment by samansmink** [comment]
LGTM!

**Comment by lnkuiper** [comment]
Thanks for the feedback @Tishj and for finding the bug @samansmink ! I have updated the PR and pulled Sam's fix into it.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**